### ISSUE TITLE
"python.analysis.diagnosticSeverityOverrides" is ignored when pyrightconfig.json is given in folder

### ISSUE THREAD

**Issue by jamilraichouni** [microsoft/pyright#836]
**Describe the bug**
When there is a folder in a workspace with a file `pyrightconfig.json` at root level of that folder that contains some valid pyright config or is even empty the following settings from `Code/User/settings.json` are ignored:

```json
{
    "python.analysis.diagnosticSeverityOverrides": {
        "reportGeneralTypeIssues": "warning",
        "reportMissingTypeStubs": "information",
        "reportUnboundVariable": "warning",
        "reportUndefinedVariable": "none",
        "reportUnknownMemberType": "warning",
        "reportUnknownVariableType": "warning",
        "reportUntypedClassDecorator": "none",
        "reportUntypedFunctionDecorator": "none",
        "reportUnusedImport": "none"
    }
}
```

**To Reproduce**
Create a folder with a sub directory `src` and a `script.py` like that:

```python
import pandas
```

Put the settings listed at the top of this report into the file `Code/User/settings.json`.

You will see the following in  the problem panel:
![expected](https://user-images.githubusercontent.com/45113557/87539293-e7d6ca80-c69d-11ea-8f55-667c8ec53f46.png)

Now create an empty file `pyrightconfig.json` and **reload the window**.
The settings for the severities are gone:
![actual](https://user-images.githubusercontent.com/45113557/87539318-f1f8c900-c69d-11ea-886b-e33c7ad47701.png)

**Expected behavior**
The problem `reportMissingTypeStubs` shall keep having a severity level of type `information` as shown on the first screenshot. It is a use case, to define severities on a global (user settings) level but also create `pyrightconfig.json` files per project/ folder to set includes, excludes etc..

The settings concept of Visual Studio Code shall be considered/ followed:

The order for over-ruling is as described in the [documentation for the WorkspaceConfiguration interface of VSCode](https://github.com/microsoft/vscode/blob/aa930e56196a1af56cae6d70b0e424fa0e99b824/src/vs/vscode.d.ts#L4674) 

**Screenshots or Code**
see above

**VS Code extension or command-line**
I'm running pyright as a VS Code extension in version 1.1.52

**Additional context**
Nothing to add.

**Comment by erictraut** [comment]
The settings precedence that you cited above is specific to VS Code settings. It doesn't apply to configuration files outside of VS Code, like pyrightconfig.json.

The reason that pyrightconfig.json overrides the VS Code settings is to guarantee consistent behavior between the command-line version of Pyright (which is often used for continuous integration pipelines) and local editing. If users are able to override these settings locally in VS Code, some developers could see diagnostic messages that are inconsistent with those reported by their CI pipeline.

Can you explain your use case in more detail? Why are you using a pyrightconfig.json file, and why do you want the VS Code settings to override it?

**Comment by jamilraichouni** [comment]
Hi Eric,

Many thanks for your quick feedback!

Understand your point. It may be useful to be able to define settings like
`"pyright.include"` and `"pyright.exclude"` with scope `"resource"` (see use case below).

# Effective value for a setting

Reading the mentioned [documentation for the WorkspaceConfiguration interface of VSCode](https://github.com/microsoft/vscode/blob/aa930e56196a1af56cae6d70b0e424fa0e99b824/src/vs/vscode.d.ts#L4674):

The *effective* value (returned by [`get`](#WorkspaceConfiguration.get) and excluding language settings) is computed by overriding or merging the values in the following order.

1. `defaultValue`
1. `globalValue` (if defined)
1. `workspaceValue` (if defined)
1. `workspaceFolderValue` (if defined)

**Note:** Only `object` value types are merged and all other value types are overridden.

# Use case

## Global severity overrides

Define severities (`"python.analysis.diagnosticSeverityOverrides"`) in a way following the hierarchy listed above.

**Example:**

Set `diagnosticSeverityOverrides` in global settings (`Code/User/settings.json`) since a user may always want to apply the same severity overrides for all projects.

## Fine-tuned `include`/ `exclude` settings for best speed of analysis

Set `"include"` and/ or `"exclude"` per workspace or folder (not only in `pyrightconfig.json` but also in VSCode's config system) since project directory names differ from project to project.

**Example:**

There might be a

* project A with a sub directory `${workspaceFolder}/pkg`,
* project B with a sub directory with `${workspaceFolder}/lib`,
* project X, Y, and Z each with a sub directory `${workspaceFolder}/backend`.



As described at the top of this message:

For all projects the user may want to have identical severity overrides that
have been defined **globally** in `Code/User/settings.json` but:

The user may also want to fine-tune include/ exclude settings (see below):

1. For the projects A and B the include/ exclude settings may be defined as
   **folder settings**.
1. For the other (backend) projects X, Y, and Z that are organised in one `.code-workspace`
   the user may want to define include/ exclude **workspace** settings.

## Folder settings for `include` and `exclude`

**Project A:**

`${workspaceFolder}/.vscode/settings.json` with something like
```json
{
    "pyright.include": ["pkg"]
}
```

**Project B:**

`${workspaceFolder}/.vscode/settings.json` with something like
```json
{
    "pyright.include": ["lib"]
}
```

## Workspace settings for `include` and `exclude`

**Projects X, Y, and Z:**

`web_applications.code-workspace` with something like
```json
{
    "folders": [
        {
            "path": "/path/to/X"
        },
        {
            "path": "/path/to/Y"
        },
        {
            "path": "/path/to/Z"
        },
    ],
    "settings": {
        "pyright.include": ["backend"]
    }
}
```

# Benefits

1. Capability to set includes/ excludes per folder without the need (just a possibility)
   to create and maintain an extra file `pyrightconfig.json`.
1. Possibility to set includes/ excludes at one location for multiple projects organised
   in one `.code-workspace`.

Hope that makes sense.

**Comment by jamilraichouni** [comment]
A solution might be, that all pyright settings can be defined within VSCode's config system and also via a `pyrightconfig.json` file for CI pipelines and/ or running the pyright CLI from a shell like the integrated terminal.

**Comment by erictraut** [comment]
Pyright offers many advanced (and rather esoteric) configuration features. We generally expose only frequently-used or highly-requested configuration features as VS Code settings. The remainder of the configuration settings are available only in the config file.

The Pyright VS Code extension will be deprecated at some point in favor of Pylance. All Pylance settings use the prefix "python.analysis.xxx", so if we were to expose the VS Code settings that you've suggested, we would do so in Pylance rather than in Pyright.

I think you've made a good case for several new settings. I recommend that you create a new feature request in the [pylance-release github repo](https://github.com/microsoft/pylance-release) and copy your use case above. If enough other users upvote the feature request, we can include it in future versions of Pylance.

In the meantime, I'm going to close this ticket because the original issue that you reported is working "as designed" in Pyright.

**Comment by jamilraichouni** [comment]
Will do so, thanks!

**Comment by jamilraichouni** [comment]
see microsoft/pylance-release#143
